Add Hostapi Pulseaudio to Portaudio

illuusio opened the pull request:
This Pull Request adds Pulseaudio Hostapi to Portaudio

illuusio committed:
Update text in README.md

illuusio committed:
Add HostAPI Pulseaudio-driver to Portaudio. Features Callback interface and Blocking implementation.
  * Pulseaudio (http://www.freedesktop.org/wiki/Software/PulseAudio/)
  * Official Portaudio (http://www.portaudio.com/)

illuusio committed:
Changes need in Portaudio to make Pulseaudio work

illuusio committed:
Code formating that is follows ANSI-coding style which seems to be Portaudio coding convention

daschuer commented:
I have failed to build this. What is the receipt? 

I have tried to call 
autoreconf -if

but its has an error and a lot of warnings: 

configure.ac:38: error: required file '../../ltmain.sh' not found

I am on Ubuntu Trusty

illuusio commented:
I use

<pre>
aclocal
libtoolize -f
autoconf -f
automake -f
</pre>

but you can also use (if above doesn't work)

<pre>
aclocal -f
libtoolize -f -c
autoconf -f
automake -f -a
</pre>

I should check the CMake version because it's should be less pain

daschuer commented:
I have still no luck: 

automake -f 
-aautomake: warning: autoconf input should be named 'configure.ac', not 'configure.in'
configure.in: error: no proper invocation of AM_INIT_AUTOMAKE was found.
configure.in: You should verify that configure.in invokes AM_INIT_AUTOMAKE,
configure.in: that aclocal.m4 is present in the top-level directory,
configure.in: and that aclocal.m4 was recently regenerated (using aclocal)
automake: error: no 'Makefile.am' found for any configure output

illuusio commented:
Oh there should be './configure' now? Just run it and it compiles. It's very frustrating that autotools configuration is messing up.

daschuer commented:
Ah thanks, I did not expect that this worked after previous error. 
Now it woks ... Great :-)

illuusio commented:
Marvellous @daschuer. I'll update 'Makefile.am' and 'README.md'. After 15 years of maintaining autotools in several projects I've have love and hate relationship with it. Glad you made it work!

daschuer commented:
What will be the best way to handle installations without Pulse? 
The ditros may provide two versions of Portaudio one with Pulse and one without. 
But this requires extra maintenance. How about dynamically load the libpulse? 
If it fails, continue without pulse api.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+/* PaPulseaudioHostApiRepresentation - host api datastructure specific to this implementation */
+
+#define PULSEAUDIO_TIME_EVENT_USEC 50000
+#define PULSEAUDIO_BUFFER_SIZE (88100 * 4 * 2)
```
Comments why you pick these values above would be helpful

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+
+fail:
+    PulseaudioFree(ptr);
+    free(ptr);
```
This looks like double free

daschuer replied:
don't use free() here, use PaUtil_FreeMemory()

daschuer replied:
Would id be possible to move the PaUtil_FreeMemory()  call to PulseaudioFree() ?

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+    char proc[PATH_MAX];
+    char buf[PATH_MAX + 20];
+
+    ptr = (PaPulseaudioHostApiRepresentation*)PaUtil_AllocateMemory(sizeof(PaPulseaudioHostApiRepresentation));
```
I think the memory is not initialized with zero, which is assumed later.

illuusio replied:
Actually it's malloc which returns NULL or pointer to memory but I also added memset to make sure we don't have problems.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+static void PulseaudioCheckContextStateCb(pa_context * c, void *userdata)
+{
+    PaPulseaudioHostApiRepresentation *ptr = userdata;
+    assert(c);
```
can we handle this situation gracefully?

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+              strnlen(l_strName, 1024) + 1), paInsufficientMemory);
+
+    memset((char *)l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].name, 0x00, strnlen(l_strName, 1024) + 1);
+    strncpy((char *)l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].name, l_strName, strnlen(l_strName, 1024));
```
I don't now the PA style guide, but these lines above are not entirely displayed in github.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+    }
+
+    PA_UNLESS(l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount] = (char*)PaUtil_GroupAllocateMemory(l_ptrHostApi->allocations,
+              strnlen(l->name, 1024) + 1), paInsufficientMemory);
```
strnlen call return value could be cached in a local var.

illuusio replied:
Don't really get this one? It's malloc group.

daschuer replied:
I meant: Use  
`size_t nameLen = strnlen(l->name, 1024);`
instead of call strnlen() twice, since it parses the string every call.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+    PA_UNLESS(l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount] = (char*)PaUtil_GroupAllocateMemory(l_ptrHostApi->allocations,
+              strnlen(l->name, 1024) + 1), paInsufficientMemory);
+
+    memset(l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount], 0x00, strnlen(l->name, 1024) + 1);
```
all zeros are overwritten by the following call except the last. 
use 
l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount][strnlen(l->name, 1024)] = '\0';

illuusio replied:
Same with this. Last character ain't zero?

daschuer replied:
Currently it is like this: 

name = "hallo\0"
strnlen(name,3) = 3; 
malloc 3 +1 -> "xxxx"
memset 4 -> "\0\0\0\0"   <- Loop overt the string 
strncpy 3 -> "hal\0"

You can replace it with the following to save the memset call 

name = "hallo\0"
strnlen(name,3) = 3; 
malloc 3 +1 -> "xxxx"
[3] = '\0' "xxx\0"  <- write only one byte 
strncpy 3 -> "hal\0"

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+              strnlen(l_strName, 1024) + 1), paInsufficientMemory);
+
+    memset((char *)l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].name, 0x00, strnlen(l_strName, 1024) + 1);
+    strncpy((char *)l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].name, l_strName, strnlen(l_strName, 1024));
```
Same as above.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+{
+    const pa_buffer_attr *a;
+    char cmt[PA_CHANNEL_MAP_SNPRINT_MAX], sst[PA_SAMPLE_SPEC_SNPRINT_MAX];
+    assert(s);
```
Can we handle this gracefully?

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+void PulseaudioStreamStateCb(pa_stream *s, void *userdata)
+{
+    const pa_buffer_attr *a;
+    char cmt[PA_CHANNEL_MAP_SNPRINT_MAX], sst[PA_SAMPLE_SPEC_SNPRINT_MAX];
```
Not used.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+
+    if (stream->underflows >= 6 && stream->latency < 2000000)
+    {
+        stream->latency = (stream->latency * 3) / 2;
```
What does this mean for Mixxx? If this has no effect for Mixxx, I am in doubt that this will improve the low CPU state much.

daschuer replied:
Do other APIs have such a feature?

illuusio replied:
Every API report latency and XRun. In this case it's not implemented correctly but this tries to play nice and not eat all your resources.

daschuer replied:
I am trying to understand, please correct me If I am wrong: 

Mixxx is driven by the Pulse callback 
An underflow happens if the callback is not returned before the Samples are required. 
This function is probably called from the same thread after Mixxx returns late.
In the meantime The soundcard plays zeros, or the old buffer again  
Pulse call the callback again, has skipped one, since the Zeros are already send. 
This means for Mixxx, which is timed by the callbacks, that the track stops for a cycle 
(bad for dancers because they fall out of rythm, but that is a separate issue) 
Mixxx continues as usual. 
The latency has not changed. 

stream->latency is used to allocate buffers. Original we had also a similar latency parameter in Mixxx, we have renamed it because the name was misleading.
Now it is the audio buffer size, which of cause effects directly the latency. 

We may increase the audio buffer size if we experience underflows, but Mixxx is not aware of that. 
Port-audio did not even know if this even helps. 
IMHO it is better lets just report underflows and let the Application or User decide, but I am still unsure if I got all right.

illuusio replied:
Actually only writing is driven by callback and reading blocking. Hmm.. how do I explain this. You are almost correct but you got it little bit wrong.
Reported latency by PulseAudio is latency from reading/writing ALSA which is reported by soundchip and kernel. PulseAudio undderrun is situation where we have too much latency (for keeping stream alive) for example we can't write fast enough to speakers.
So it's misleading because you can only estimate the latency from mixing to PulseAudio to ALSA to Chip to Speakers (Actually PulseAudio is as fast as Jack but that is another story). If there is no underrun we just keep making CB and dancers are :dancers:.
PulseAudio does callbacks even if we don't feed anything out or in then we have undderun. It tries to keep you up and running in selected sample rate and sample mode (S16, Float). What we feed through the Portaudio API is just as fast as we get it through PulseAudio CB and then to the Mixxx or Pulseaudio asks for samples through Portaudio (As ALSA does) and then it feeds them to ALSA.
You should refer this one: https://github.com/illuusio/linux-audio-example where I tried to make simple examples how different Linux Audio APIs work. It's work in process but contain Pulseaudio and Portaudio so you can see how they should work.
And yes with bigger buffer size we can tolare undderrun better but if it's happen then data is lost for ever.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+    *hostApi = &l_ptrPulseaudioHostApi->inheritedHostApiRep;
+    (*hostApi)->info.structVersion = 1;
+    (*hostApi)->info.type = paPulseaudio;
+    (*hostApi)->info.name = "Pulseaudio";
```
"PulseAudio" is always written with camel notation

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+    (*hostApi)->info.defaultOutputDevice = paNoDevice;
+
+    /* Connect to server */
+    pa_context_connect(l_ptrPulseaudioHostApi->context, NULL, 0, NULL);
```
return value is ignored.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+    /* Connect to server */
+    pa_context_connect(l_ptrPulseaudioHostApi->context, NULL, 0, NULL);
+
+    while(1)
```
This loop should be replaced, by signal handling using the state callback.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+/* Allocate buffer. */
+static PaError PulseaudioBlockingInitRingBuffer(PaPulseaudioStream *stream, PaUtilRingBuffer *rbuf, int size)
+{
+    long l_lNumBytes = 4096 * size;
```
Why 4096?

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+    }
+
+    memset(buf, 0x00, PATH_MAX + 20);
+    snprintf(buf, sizeof(buf), "Portaudio Pulseaudio HostApi");
```
It would be better to use her the client application name. 
That can be read from for example form 
/proc/4475/cmdline  (4475 = getpid())
or other hacks.

daschuer replied:
This works for me (and only with Linux/gcc/glibc):

<pre>
#include &lt;stdio.h&gt;

extern char *program_invocation_name;
extern char *program_invocation_short_name;
extern char *__progname;

void main()
{
   printf("%s\n",program_invocation_short_name);
   printf("%s\n",program_invocation_name);
   printf("%s\n",__progname);
}
</pre>

_Updated full example if someone reads this and wonders how it really works_

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+            goto error;
+        }
+
+        stream->inStream = pa_stream_new(l_ptrPulseaudioHostApi->context, "Portaudio source", &stream->inSampleSpec, NULL);
```
Use here the application name as well

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+
+    pa_threaded_mainloop_lock(l_ptrPulseaudioHostApi->mainloop);
+
+    if(pa_stream_get_time(stream->outStream, &stream->outStreamTime) !=  -PA_ERR_NODATA)
```
invert if

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.h:
```diff
@@ -0,0 +1,155 @@
+
+
+#define PULSEAUDIO_TIME_EVENT_USEC 50000
+#define PULSEAUDIO_BUFFER_SIZE (88100 * 4 * 2)
```
redifined in C source

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c:
```diff
@@ -0,0 +1,483 @@
+    /* Ready the processor */
+    PaUtil_ResetBufferProcessor(&stream->bufferProcessor);
+
+    stream->latency = 20000;
```
Why is this a constant?

illuusio replied:
I don't really remember. I think it felt nice in the days I implemented this :rabbit:

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c:
```diff
@@ -0,0 +1,483 @@
+
+            pa_threaded_mainloop_unlock(l_ptrPulseaudioHostApi->mainloop);
+
+            usleep(1000);
```
Can we use signals here?

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c:
```diff
@@ -0,0 +1,483 @@
+    int i = 0;
+
+    assert(s);
+    assert(length > 0);
```
Handle gracefully

illuusio replied:
Hmm.. have any tip for that? This is from Portaudio skeleton.

daschuer replied:
Ah you are right, this is the way Portaudio is designed. So keep it, we have to only be sure, that it catches only coding issues and does not crash because of any runtime exception on the users machine.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c:
```diff
@@ -0,0 +1,483 @@
+    pa_threaded_mainloop_signal(l_ptrStream->mainloop, 0);
+}
+
+void PulseaudioStreamWriteCb(pa_stream *s, size_t length, void *userdata)
```
When is this called an from where? Some comment could be help here and elsewhere. 
Since this code is between the application and Pulse, it is sometimes confusing. 
Especially because of the common pa prefix.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c:
```diff
@@ -0,0 +1,483 @@
+    const void *l_ptrSampleData = NULL;
+
+    assert(s);
+    assert(length > 0);
```
why is length unused in the following code?

illuusio replied:
Just make this not to mention unused variables. Actually this is the magic and problem with PulseAudio. You just 'pa_stream_peek' and you get as much as there is. You can read the length but it's safer to just read everything and hope for the best.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1041 @@
+
+    for(i = 0; i < 1024; i ++)
+    {
+        memset(l_ptrPulseaudioHostApi->deviceInfoArray, 0x00, sizeof(PaDeviceInfo) * 1024);
```
can this be done outside the loop?

daschuer commented:
Fist reading path finished. 
Unfortunately I did not understand the whole yet. 
I hope this will change during the review process. 
In any case this would be a real good addition to Mixxx.

illuusio committed:
Use variable '__programname' (which contains current running binary name) as PulseAudio client name

illuusio committed:
Remove double introduction of defines as they are in header file also and not used in this file

illuusio committed:
Remove double introduction of 'PA_PULSEAUDIO_SET_LAST_HOST_ERROR' it is in header file

illuusio committed:
Added comments to defined values

illuusio committed:
Added comments what happens inside function: 'PulseaudioNew'. Added HOST error messages

illuusio committed:
Be sure that we don't have double free problems anymore by using 'PulseaudioFree' which frees PulseAudio and HostApi and not using 'free' to favor 'PaUtil_FreeMemory'

daschuer commented on configure.in:
```diff
@@ -502,6 +520,7 @@ case "$target_os" in
 	AC_MSG_RESULT([
   OSS ......................... $have_oss
   JACK ........................ $have_jack
+  Pulseaudio .................. $have_pulse
```
PulseAudio

illuusio committed:
Gracefully handle asserts and add them to all PulseAudio callbacks

illuusio committed:
Commnet out unused variable which are used in debug

illuusio committed:
Big rename from 'Pulseaudio' to 'PulseAudio' which is Official way of writing PulseAudio

illuusio committed:
Update 'Pulseaudio' to official 'PulseAudio'

illuusio commented:
Lazy day and <code>sed -i</code> was my friend to change Pulseaudio to PulseAudio. Can't grep any Pulseaudio anymore but there can be typos is writing but they'll be founded soon enough. All the init code which I have written 2 years ago seems to be little bit rusty :-1:

illuusio committed:
Cut lines at 80 char and streamline code indenting all the way

illuusio committed:
Clean-up #define that is declared in header

illuusio commented:
Ok now all the things should be 'solved' or commented. I have to work with Init code as I said it's rusty and I made it before I understood nothing about Portaudio or PulseAudio.
How to work with just ALSA systems. If you have compiled PulseAudio in without lazy linking (which Portaudio doesn't seem to use) you are out of luck. We just can act like Jack and make sure is we don't get into way anf gracefully drop-out.

illuusio committed:
Reworked Source and Sink list adding to one function for less copy and paste problems

illuusio commented:
Now all the things should we solved. One thing I noticed it sometimes Sigsegvs on Mixxx quit so that has to been worked out and please pay attention if it does that to you also.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1249 @@
+
+    l_ptrName = (char *)hostapi->deviceInfoArray[hostapi->deviceCount].name + l_iDeviceNameLen;
+    l_ptrName = '\0';
+    hostapi->pulseaudioDeviceNames[hostapi->deviceCount][l_iRealNameLen] = '\0';
```
[l_iRealNameLen - 1]

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1249 @@
+    strncpy((char *) hostapi->deviceInfoArray[hostapi->deviceCount].name,
+          interfaceName, l_iDeviceNameLen);
+
+    l_ptrName = (char *)hostapi->deviceInfoArray[hostapi->deviceCount].name + l_iDeviceNameLen;
```
The same here: l_iDeviceNameLen - 1

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1249 @@
+                                                  l_iDeviceNameLen),
+              paInsufficientMemory); */
+
+    strncpy((char *)
```
This cast is redundant.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1249 @@
+            hostapi->pulseaudioDeviceNames[hostapi->deviceCount],
+            realName, l_iRealNameLen);
+
+    strncpy((char *) hostapi->deviceInfoArray[hostapi->deviceCount].name,
```
name is "const char *name" such a remove const cast is ugly. You should save the allocated memory in a local. 

char *name = (char *) PaUtil_GroupAllocateMemory ...
strncpy(name, interfaceName ...
hostapi->deviceInfoArray[hostapi->deviceCount].name = name;

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1249 @@
+    if (l->description != NULL)
+    {
+        l_strName = (char *) l->description;
+    }
```
User here the parameter names and types of  _PulseAudioAddAudioDevice()
const char *interfaceName
const char *realName

By the way are these names reasonable, they are not uses in the Pa and not in the Pulse API.

daschuer commented on src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c:
```diff
@@ -0,0 +1,1249 @@
+    if (l->description != NULL)
+    {
+        l_strName = (char *) l->description;
+    }
```
The same here like above. Avoid const remove casts.

illuusio committed:
Correct length of string is (string + 1) - 1

illuusio committed:
Get rid of type casts

illuusio committed:
Get rid of commented out code as it's not coming back

illuusio committed:
Rename function parameters to fit more what they are

illuusio committed:
Get rid of unnecessery type casts from 'const char *' to 'char *'

illuusio committed:
Remove crust from template

illuusio commented:
Have you got good view about how Pulseaudio and Portaudio work together?

daschuer commented:
I have to test the latest changes. But the over all impression is quite good. I would still directly use ALSA for pro setups, but for my home jukebox setup it seems to work better than the pulse device we have already

illuusio commented:
Nice to hear it at least works. I think I'll announce this on Portaudio list so we can get more eyes looking at code.

Be-ing commented:
> What will be the best way to handle installations without Pulse?
> The ditros may provide two versions of Portaudio one with Pulse and one without.
> But this requires extra maintenance. How about dynamically load the libpulse?
> If it fails, continue without pulse api. 

Does this actually need PA installed or just the PA libraries? Fedora and probably other distros package the PA libraries separately from PA for this reason.

daschuer commented:
Others than ALSA, we cannot assume that  Pulse is installed. I do not like the idea that after this is merged Portaudio depends on Pulse. This would finally mean that we drew Pulse into a non pulse distro only because you install an app that requires Portaudio.  
A solution would be to provide two Portaudio version or link to pulse at runtime.

Be-ing commented:
I have no comment regarding runtime linking, but having two versions of the same package seems like a bad idea. The reason for packaging PulseAudio's libraries separately is so that programs that can use PulseAudio but don't require it can be built and installed without pulling in the dependencies of the pulseaudio executable, so the system isn't configured to use PulseAudio unless that is desired.

illuusio commented:
Let's be honest. If you are not using Slackware you are having at least libpulse installed and if you don't want that you can compile it by yourself. Fedora, Ubuntu, Debian and openSUSE at least ship it default and most of us just don't care about it.

daschuer commented:
Yes you are probably right. I was thinking about embedded systems which may not have the pulse libs. But in that case they can recompile portaudio ... on the other hand is that finally the two packages approach.

daschuer commented:
I have just had a look to the Alsa implementation and even this much more common API has an option for dynamic loading:  
https://github.com/illuusio/portaudio-pulseaudio/blob/e047b45942c1c3ad7529363ccd9b9a7414bfc82c/src/hostapi/alsa/pa_linux_alsa.c#L99
So it looks like dynamic loading belongs to a complete solution.

illuusio commented:
There is always dlopen (but it's not probably what we want) for things like this but I have to look Jack which nearly the same case. I didn't really get how that is supposed to work. They tydedef API functions and if they are available they magical link to those?

daschuer commented:
> They tydedef API functions and if they are available they magical link to those?

Yes, exactly. But lets try to remove the magic ;-) 

dlopen() loads any *.so file at runtime and does not crash if it fails. This is different to the current implementation where the staically linked *.so files are loaded at startup and crash the process if missing. 

All included functions are listed by their name (symbol) and their address. Now you can use dlsym() to lookup the function address (the function pointer) and call the function by the pointer. 
This requires that you know the parameter set before, there is no check that you call the function with the correct parameter. 

Hope that helps.

illuusio committed:
Make configure '--with-pulseaudio=yes --with-jack=no' work by adding 'PKG_PROG_PKG_CONFIG' (which added only when it's needed)
Add 'pa_ringbuffer.o' when it's needed by Jack or PulseAudio.

illuusio committed:
Re-name all 'pulse/PULSE' to pulseaudio/PULSEAUDIO

illuusio committed:
Now when cannot connect to PulseAudio server then handle it gracefully and let others HostApis to handle situation

illuusio commented:
Thanks for explanation @daschuer. I'll try to find time for the dynamic loading. I just fixed autoconf and configure.in (I think is lose rest of my hear but :haircut: is still on and things should be ok). I don't know if scons sometimes feels like pain autoconf is just torture. There also should be fixed case when there ain't server.

illuusio committed:
Fix Jack/Pulseaudio findindn when no commandline options are given to 'configure.in'

illuusio commented:
@sqweek can you try this basic example:
https://github.com/illuusio/linux-audio-example/tree/master/portaudio
Do you still hear noise in start of blocking stream? You have Jack all the time running or just when you need it? Is Pulseaudio routed through ALSA or Jack? Just asking because I can't hear any noise starting blocking stream.

illuusio committed:
Changed waiting got connection simpler and let PulseAudio make work for us.

illuusio committed:
We need to make zeroing 'deviceInfoArray' only once

illuusio committed:
Less locking because these should be thread safe and less locking is allways better

illuusio committed:
Added some comments to make more clear what these are doing

sqweek commented:
Hi Tuukka, hope your holidays have been good! :)
I tried your linux-audio-example and can't hear any glitches there. Well, I heard one glitch around EOF but that's because the code sends the full buffer even if the read didn't fill it.

So it would seem that the glitch I'm hearing is caused by either the golang bindings I'm using, or more likely my application itself. I edited libsndfile_port_blockplay.c to test a few things but even after ensuring it was using pulse (I think the default device is via ALSA on my system), matching sample formats and stream flags, and inserting a sleep in between starting the stream and writing samples, I couldn't cause any glitches.

Thanks for the test case and sorry to take up your time. To answer your other questions I have pulse routed through ALSA, and normally jack isn't running (I start it manually when I get sick of pulse :P).

I've started looking at the libpulse code to diagnose the crash but haven't come to a conclusion yet. So far it looks more like an internal logic error in the libpulse mainloop/client code rather than incorrect usage on the portaudio side.

illuusio commented:
Holidays have been very nice and uplifting. I have to investigate little bit Go bindings :mag:. If they are making some strange things or there is some bugs that I haven't found (something that is not compatible with ALSA). I wanted to investigate Go but have no excuse for it until now..
Hmm.. I don't know if it's a suicide to go to PulseAudio devel mailing list and ask? They are little bit like Kernel mailing list.. not very polite. Please test this new version of waiting. Is it fixing things for you.

sqweek commented:
There is something very strange going on - even after my app stops playback (via Pa_AbortStream) I sometimes hear crackling noises for a few seconds. I'll see if I can reproduce with a minimal Go program, but not tonight as I need sleep :zzz:. The bindings I'm using are https://github.com/gordonklaus/portaudio - apparently they aren't working in Go 1.6beta1, I'm on Go 1.5.2.

daschuer commented:
I have tested this branch with Ultrastar Delux a Karaoke application, but it is crashing. 
After restoring the original branch it works again.

sqweek commented:
Aha, figured it out. Was nothing to do with Go in the end - the difference is that my program uses much smaller buffers with Pa_WriteStream. I managed to reproduce glitches with the following patch to linux-audio-example:

```
diff --git a/portaudio/libsndfile_port_blockplay.c b/portaudio/libsndfile_port_blockplay.c
index a6100a0..d3debeb 100644
--- a/portaudio/libsndfile_port_blockplay.c
+++ b/portaudio/libsndfile_port_blockplay.c
@@ -49,7 +49,7 @@
 SNDFILE *infile;
 SF_INFO sfinfo ;

-#define PLAY_FRAMES_PER_BUFFER 44100
+#define PLAY_FRAMES_PER_BUFFER 1024

 /* Handle termination with CTRL-C */
 static void handler(int sig, siginfo_t *si, void *unused) {
@@ -102,7 +102,8 @@ int main(int argc, char *argv[]) {
         goto exit;
     }

-    outputParameters.device = Pa_GetDefaultOutputDevice(); /* default output device */
+    outputParameters.device = Pa_HostApiDeviceIndexToDeviceIndex(Pa_HostApiTypeIdToHostApiIndex(paPulseAudio), 0);

     if (outputParameters.device == paNoDevice) {
         fprintf(stderr, "Error: No default output device.\n");
```

I think it only affects the first call to Pa_WriteStream (each time after calling Pa_StartStream).

sqweek commented:
@daschuer Does it crash on startup, or when playback starts, or something else? Do you get one of these failures in the console?

```
Assertion 'c->callback' failed at pulsecore/socket-client.c:126, function do_call(). Aborting.
Assertion 'c->defer_event == e' failed at pulsecore/socket-client.c:170, function connect_defer_cb(). Aborting.
```

illuusio commented:
Mixxx is behaving pretty nice when opening devices. I test if Ultrastar deluxe also crashes for me and then I get bottom of the problem (Nice to see some major Pascal coding again!). @sqweek I think I nailed down last night (before sleep) what is causing the click and I have to investigate if I'm correct.
I think problem is;

<pre>
Assertion 'o' failed at pulse/operation.c:133, function pa_operation_get_state(). Aborting.
</pre>

<code>
#0  0x00007ffff45cc187 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff45cd538 in __GI_abort () at abort.c:78
#2  0x00007ffff3b7c0d2 in pa_operation_get_state () from /usr/lib64/libpulse.so.0
#3  0x00007ffff6e16e04 in PulseAudioCloseStreamCb (s=0x8bfbe0) at src/hostapi/pulseaudio/pa_hostapi_pulseaudio_cb.c:323
#4  0x00000000004d8a6f in UAUDIOINPUT_PORTAUDIO_TAUDIOINPUT_PORTAUDIO_$__ENUMDEVICES$$BOOLEAN ()
#5  0x00000000004d8e43 in UAUDIOINPUT_PORTAUDIO_TAUDIOINPUT_PORTAUDIO_$__INITIALIZERECORD$$BOOLEAN ()
#6  0x0000000000000001 in ?? ()
#7  0x4ab22a0ca5c8da92 in ?? ()
#8  0x0000000000000001 in ?? ()
#9  0x00000000004cacc6 in UMUSIC_INITIALIZESOUND ()
#10 0x0000000000472694 in UMAIN_MAIN ()
#11 0x0000000000406a6f in main ()
</code>

illuusio committed:
Remove more copy'n'paste to favor one unified function

illuusio committed:
Some applications use CloseStream to iterate usable soundcards

illuusio commented:
@daschuer latest commit should make Ultrastar Deluxe work. Happy singing with PulseAudio..

sqweek committed:
pulse: Fix crash during PaPulseAudio_Initialize

When using a threaded-mainloop, pulseaudio calls require
synchronisation via pa_threaded_mainloop_lock and _unlock.

This fixes errors of the form:

 Assertion 'c->callback' failed at pulsecore/socket-client.c:126, function do_call(). Aborting.
 Assertion 'c->defer_event == e' failed at pulsecore/socket-client.c:170, function connect_defer_cb(). Aborting.

sqweek commented:
@illuusio I fixed the crash during startup! I don't fully understand the race but this was sufficient: 36a745d7a3a

libpulse doesn't seem to use `pa_socket_client.defer_event` outside connection which is what led me to try this. But the field does get cleared after every `do_call()` so a connection attempt while some other operation is in progress would be enough. No clue what else pulse might be doing though...

illuusio commented:
@sqweek wonderful! Ok now I understand the problem. I have to test can we start mainloop after connection (which is the 'correct' fix. If not that one is what works best). Those lock and unlock locks threaded mainloop and it seems that it is not thread safe. BTW. I'm little bit confused where is that commit?

sqweek commented:
Usually with git everyone works with their own copy of the repo, also called a 'fork'. I made a fork of yours (you can see it at http://github.com/sqweek/portaudio-pulseaudio) which is where I pushed the commit. If you like I can make a pull request (which makes it easy for you to integrate my commit into your fork); I held off because I wasn't sure if you had unpushed commits and don't want to disrupt your workflow.

edit: I have to admit I'm not 100% sure how github's SHA1 hyperlinking magic works. I thought I'd have to link directly to my fork but it turned out the SHA1 was enough by itself.

edit 2: sorry if the above came off as patronising, i'm not sure if i've confused you with someone else on the mailing list who recently said they weren't very familiar with git.

illuusio commented:
I'm basicly old CVS/Subversion fart (trying to learn Git and Mercurial) and I just don't understand all the :octocat: goodness (I though you have forged this repo but GUI showed that it was against my repo). I think we have to wrap all the places where context is used with locking which it's just kind of funny  because we make context out of threaded-mainloop. If you do that and make PR then I can try to merge it.

illuusio committed:
Merge pull request #2 from sqweek/hostapi-pulseaudio

Fix PulseAudio threading mainloop crashers with adding needed locking for resources that are not thread safe.

sqweek commented:
@illuusio Did you get anywhere investigating the click? I think the problem is that `PulseAudioWriteStreamBlock` assumes the provided buffer is at least as long as the remaining space in the ring buffer. So writing a small chunk of audio when the ring buffer is empty ends up filling the ring buffer using memory past the end of the provided audio data.

Also I don't think calling `PaUtil_GetRingBufferWriteAvailable` multiple times is a safe approach, as the ring buffer's available space could change at any time (depending on when the audio thread reads a chunk). It could even change between GetWriteAvailable and WriteRingbuffer. This could also lead to glitches by repeating parts of the audio stream.

Maybe this is prevented by taking the pulse-audio lock (I haven't figured out exactly what is happening on the other side of the ring buffer), but if so that means we're blocking the audio thread which is not a good idea. It can also cause glitches if the kernel doesn't reschedule the audio thread in time to finish processing the current buffer before the sound card needs it.

I think it would be simpler and safer to just call `PaUtil_WriteRingBuffer` in a loop with the number of bytes remaining in the audio data, and update `l_lLength` and `l_ptrData` based on the return value of `WriteRingBuffer` (ie. the number of bytes actually written). This should also allow the lock to be removed safely, as the whole point of portaudio's ring buffer is to provide lockless communication between threads.

Does that make sense? I'm happy to prepare a PR if you are sick of looking at this code :)

illuusio commented:
Yes you are correct @sqweek. Click or noise is coming unfilled buffer. I've been busy to fix this but I have been thinking to fix it with setting PulseAudio as cork (IDLE) as first and then start feeding after we have enough stuff in ring buffer just start feeding. You are more than welcome to make PR and I'll make comments and commits to it if there is something I like to change and then merge it.

sqweek committed:
pulseaudio: simplify input ring-buffer handling

illuusio committed:
Make Portaudio Terminate-callback work as it should. Mainly Add more locking for prevent crashes and faster exit
also fix logic of disconnection.

illuusio committed:
Output stream Stopping and starting is working as expected.

illuusio committed:
Fix Portaudio block playing API. PulseAudio is not supporting blocking writing but provides 'pa_stream_writable_size'
for quering how much can we write. So with turning OFF PulseAudio write-callback and just looping until we have
written everything emulates block writing very well.

illuusio commented:
Now there is some callback clicks to investigate (which are caused same that caused the blocking playback clicks) and stopping and starting input stream but otherwise this start to be in very good shape.. Hopefully I didn't mugged up Terminate with last fix :facepunch:

sqweek commented:
Hm. My program did have a click in callback mode, which I managed to fix by changing my callback to zero the entire buffer -- before that if I didn't have a full buffer's worth of samples I only filled in part of the buffer. Don't know if that helps.

There's a few more problems I'd like to investigate, how would you prefer to manage them? I can make a PR for each, or open an issue, or just make a list here?

illuusio commented:
PR is most convenient I think. List what is wrong could be also good so I can fix some of them also.

sqweek commented:
Actually the only definite problem I'm currenty aware of is that the left/right channel balance is off when the program opens a stream in mono mode (eg. patest_mono). It should be centered but it is heavily skewed towards one of the channels.

(the other things on my list I need to investigate to see if there is actually a problem before bothering you)

sqweek committed:
pulse: fix error handling in OpenStream

1. Change PulseAudioConvertPortaudioFormatToPulseAudio to return
paSampleFormatNotSupported on failure instead of paNotInitialized.

2. Fix OpenStream to check the return value from:
  * PulseAudioConvertPortaudioFormatToPulseAudio
  * PulseAudioBlockingInitRingBuffer

3. Fix OpenStream to return the actual PaError on failure, instead
of always returning paNotInitialized.

4. Fix OpenStream to not free the host api representation on error
(which resulted in an assertion failure later in Terminate).

sqweek committed:
pulse: support non-interleaved audio samples

sqweek committed:
pulse: add extra braces

sqweek committed:
pulse: fix comment style

sqweek commented:
> stream in mono mode (eg. patest_mono). It should be centered but it is heavily skewed towards one of the channels.

Ok I thought portaudio's buffer processor took care of this kind of channel adaption, but it turns out that's host api specific. eg. alsa has `PaAlsaStreamComponent_DoChannelAdaption` to copy a user mono stream onto a stereo host stream.

So maybe there's nothing to do there, except document  the current behaviour (or run it by the main portaudio devs).

illuusio committed:
Merge pull request #5 from sqweek/non-interleaved

Add support for non-interleaved audio

illuusio commented:
So they do it 'hard way' I try to find time to implement it. It's not a big thing but import for many cases.

sqweek commented:
Ah I found another issue - `PulseAudioGetStreamWriteAvailableBlock` is inaccurate because it still looks at the ring buffer, which of course is no longer used for blocking output streams. Probably all references to `outputRing` can be removed?

illuusio commented:
Yes output ring buffer should be removed altogether.. probably same goes with input ringbuffer (requires some work). Can you provide some example how it's inaccurate or how to test it? I try to fix mono thing and this before eastern.

sqweek commented:
Well right now it always returns zero, but that's also because it is calling `PaUtil_GetRingBuffer*Read*Available` :). I'm not sure about a test case, but I would guess that pulseaudio has some kind of "how many samples can be written without blocking" function that could be used here?

I think the input ring buffer is still used in blocking mode at the moment? The recording side of things may need more testing in general, I know I haven't looked at it.

sqweek commented:
I ran into this:

```
Assertion 'o' failed at pulse/operation.c:133, function pa_operation_get_state(). Aborting.
```

Caused by:
1. Starting an output stream in blocking mode
2. Calling `Pa_Terminate`
3. Calling `Pa_WriteStream`

Obviously this order of operations will never work so I don't consider this a serious problem, just wanted to document it. In an ideal world it would return `paNotInitialized` instead of crashing.

illuusio commented:
Hmm.. above should be tracked as Issue? Seems to be some kind of blocker.

illuusio committed:
Convert MONO samples to stereo by copying them to temp array

illuusio committed:
More universal version of MONO to Stereo converter

illuusio committed:
Merged upstream fixes

illuusio commented:
Huh long time no pushed this but now Mono should be converted to Stereo correctly. Does this fix your problem @sqweek? I'll try to fix some input issues that I have found next and that close one too.

illuusio commented:
I'm trying to get this in officialy.. pull request to new shiny Portaudio GIT repo https://app.assembla.com/spaces/portaudio/git/merge_requests/3719213

Be-ing commented:
Awesome. Between this and the library GUI redesign, Mixxx could replace Amarok as my casual music player :)

sqweek commented:
Nice work Tuukka! :+1:

daschuer commented:
It is really bad that this is still not merged for long. Can you update this with the latest changes on assembla. I am considering to add this to a Launchpad PPA.

illuusio commented:
I've been toying around to rewrite this. I'll find time to update this

illuusio commented:
Please refer to: https://github.com/illuusio/assembla-mirror-portaudio-pulseaudio as upstream for ths
